WcaOnRails/app/webpacker/components/CompetitionOverview/CompetitionFilters.js

```diff
@@ -0,0 +1,591 @@
+} from '../../lib/wca-data.js.erb';
+
+import CompetitionTable from './CompetitionTable';
+import CompetitionMap from './CompetitionMap';
+
+import 'react-datepicker/dist/react-datepicker.css';
+
+// Max number of competitions fetched per query
+const COMPETITIONS_API_PAGINATION = 25;
+
+// Limit number of markers on map, especially for "All Past Competitions"
+const MAP_DISPLAY_LIMIT = 500;
+
+const WCA_EVENT_IDS = Object.values(events.official).map((e) => e.id);
+
+const PAST_YEARS_WITH_COMPETITIONS = [];
+for (let year = new Date().getFullYear(); year >= 2003; year -= 1) {
+  PAST_YEARS_WITH_COMPETITIONS.push(year);
+}
+PAST_YEARS_WITH_COMPETITIONS.push(1982);
+
+const regionsOptions = [
+  { key: 'all', text: I18n.t('common.all_regions'), value: 'all_regions' },
+  {
+    key: 'continents_header', value: '', disabled: true, content: <Header content={I18n.t('common.continent')} size="small" style={{ textAlign: 'center' }} />,
+  },
+  ...(Object.values(continents.real).map((continent) => (
+    { key: continent.id, text: continent.name, value: continent.id }
+  ))),
+  {
+    key: 'countries_header', value: '', disabled: true, content: <Header content={I18n.t('common.country')} size="small" style={{ textAlign: 'center' }} />,
+  },
+  ...(Object.values(countries.real).map((country) => (
+    { key: country.id, text: country.name, value: country.iso2 }
+  ))),
+];
+
+const filterInitialState = {
+  timeOrder: 'present',
+  selectedYear: 'all_years',
+  customStartDate: null,
+  customEndDate: null,
+  region: 'all_regions',
+  delegate: '',
+  search: '',
+};
+
+const calculateQueryKey = (filterState) => {
+  let timeKey = '';
+  if (filterState?.timeOrder === 'past') {
+    timeKey = `${filterState.selectedYear}`;
+  } else if (filterState?.timeOrder === 'custom') {
+    timeKey = `start${filterState.customStartDate}-end${filterState.customEndDate}`;
+  }
+
+  return {
+    timeOrder: filterState?.timeOrder,
+    region: filterState?.region,
+    delegate: filterState?.delegate,
+    search: filterState?.search,
+    time: timeKey,
+  };
+};
+
+function CompetitionFilters() {
+  const [selectedEvents, setSelectedEvents] = useState([]);
+  const [shouldShowRegStatus, setShouldShowRegStatus] = useState(false);
+  const [shouldShowCancelled, setShouldShowCancelled] = useState(false);
+  const [displayMode, setDisplayMode] = useState('list');
+  const [competitionData, setCompetitionData] = useState([]);
+  const [competitionQueryKey, setCompetitionQueryKey] = useState({
+    timeOrder: 'present', region: 'all_regions', delegate: '', search: '', time: '',
+  });
+
+  const editSelectedEvents = (eventId) => {
+    setSelectedEvents((prevSelectedEvents) => (
+      prevSelectedEvents.includes(eventId)
+        ? prevSelectedEvents.filter((id) => id !== eventId)
+        : [...prevSelectedEvents, eventId]
+    ));
+  };
+
+  const filterReducer = (state, action) => {
+    const newState = { ...state, ...action };
```

Comment:
Typically the reducer `action` is an object with a `type` property along with whatever data is required to perform that action. Such as `{ type: 'update-region', region: [...] }`, `{ type: 'search', text: [...] }`, or `{ type: 'reset' }`. I'm not sure it's necessarily wrong to just pass in the partial object and merge that into the existing object though.

Reply:
I looked up examples of `useReducer` and was going to do that, but I was unsure if that would be redundant.